packages/flutter/lib/src/widgets/router.dart

```diff
@@ -998,6 +998,63 @@ class ChildBackButtonDispatcher extends BackButtonDispatcher {
   }
 }
 
+/// A convenience widget that registers a callback when the back button is pressed
+/// by using [BackButtonDispatcher].
+///
+/// It only applies to platforms that accept back button clicks, such as Android.
+///
+/// It can be useful for scenarios, in which you create a different state in your
+/// screen but don't want to use a new page for that.
+class BackButtonListener extends StatefulWidget {
+  /// Creates a BackButtonListener widget .
+  ///
+  /// The [child] and [onBackPressed] arguments must not be null.
+  const BackButtonListener({
+    Key? key,
+    required this.child,
+    required this.onBackPressed,
+  }) : super(key: key);
+
+  /// The child Widget that will be drawn and usually a page that will
+  final Widget child;
+
+  /// The callback function that will be called when the back button is pressed.
+  /// It must return a boolean future with true if this child will handle the request;
+  /// otherwise, return a boolean future with false.
+  final ValueGetter<Future<bool>> onBackPressed;
+
+  @override
+  _BackButtonListenerState createState() => _BackButtonListenerState();
+}
+
+class _BackButtonListenerState extends State<BackButtonListener> {
```

Comment:
We also need to implement didUpdateWidget, to handle the case where user swapped out onBackPressed. In that case, we need to reregister the call back to the dispatcher

Reply:
It will be better if we also have a test to verify that.

Reply:
Thanks for the comments, @chunhtai. Regarding the tests, would you have any example I could look to trigger these state changes, such as `dispose` or `didUpdateWidget`?

Reply:
You can check out this example
https://gist.github.com/chunhtai/b666b3192e2356d315f9cb434d780811

Reply:
Thank you! I got it :)